lib/sgtn-client/api/translation.rb

```diff
@@ -1,81 +1,147 @@
-        s = s % args
+        s
       end
-      s
-    end
 
-    def self.getStrings(component, locale)
-      SgtnClient.logger.debug "[Translation][getStrings]component=#{component}, locale=#{locale}"
-      locale = SgtnClient::LocaleUtil.get_best_locale(locale)
-      items = get_cs(component, locale)
-      if items.nil? && !LocaleUtil.is_source_locale(locale)
-        items = get_cs(component, LocaleUtil.get_source_locale)
-        locale = LocaleUtil.get_source_locale
+      # <b>DEPRECATED:</b> Please use <tt>Singleton:get_translations</tt> instead.
+      def getStrings(component, locale)
+        warn '[DEPRECATION] `getStrings` is deprecated.  Please use `Singleton:get_translations` instead.'
+        SgtnClient.logger.debug "[Translation][getStrings]component=#{component}, locale=#{locale}"
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
+        items = get_cs(component, locale)
+        if items.nil? && !LocaleUtil.is_source_locale(locale)
+          items = get_cs(component, LocaleUtil.get_source_locale)
+          locale = LocaleUtil.get_source_locale
+        end
+
+        { 'component' => component, 'locale' => locale, 'messages' => items || {} } if items
       end
 
-      { 'component' => component, 'locale' => locale, 'messages' => items || {} } if items
-    end
+      def translate(key, component, locale: nil, **kwargs, &block)
+        translate!(key, component, locale: locale, **kwargs, &block)
+      rescue StandardError => e
+        SgtnClient.logger.error "couldn't translate #{key}, #{component}, #{locale}, #{kwargs}. error: #{e}"
+        key
+      end
+      alias t translate
 
-    def self.getTranslation(component, key, locale)
-      locale = SgtnClient::LocaleUtil.get_best_locale(locale)
-      items = get_cs(component, locale)
-      items&.fetch(key, nil)
-    end
+      # raise error when translation is not found
+      def translate!(key, component, locale: nil, **kwargs)
+        SgtnClient.logger.debug "[#{method(__callee__).owner}.#{__callee__}] key: #{key}, component: #{component}, locale: #{locale}, args: #{kwargs}"
+
+        locale = locale.nil? ? self.locale : SgtnClient::LocaleUtil.get_best_locale(locale)
 
-    def self.get_cs(component, locale)
-      id = SgtnClient::Common::BundleID.new(component, locale)
-      available_bundles = SgtnClient::Config.available_bundles
-      unless available_bundles.nil? || available_bundles.empty? || available_bundles.include?(id)
-        SgtnClient.logger.warn "[#{method(__callee__).owner}.#{__callee__}] get an unavailable bundle: #{id}"
-        return nil
+        result = get_bundle(component, locale)&.fetch(key.to_s, nil)
+        result = yield if result.nil? && block_given?
+        raise SgtnClient::SingletonError, 'translation is missing' if result.nil?
+
+        if kwargs.empty?
+          result
+        else
+          locale = result.locale if result.is_a?(SgtnClient::StringUtil)
+          result.localize(locale) % kwargs
```

Comment:
Usually `get_translations` is for frontend. So far we haven't meet strings which need pluralization.
This can be considered in future. I think other clients have the same issue.